File: src/app/product/generate-static-html.js
const fs = require('fs');
const path = require('path');

const DOMAIN = 'https://favoritelectronics.com/';

// Helper to escape HTML
function escapeHtml(str) {
  return String(str)
    .replace(/&/g, '&amp;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
}

// Helper to sanitize filename for Windows
function sanitizeFilename(filename) {
  return filename
    .replace(/[<>:"/\\|?*]/g, '') // Remove Windows forbidden characters
    .replace(/[čćđšžČĆĐŠŽ]/g, '') // Remove special characters
    .replace(/\s+/g, '-') // Replace spaces with hyphens
    .replace(/[^\w\-\.]/g, '') // Keep only alphanumeric, hyphens, and dots
    .substring(0, 200); // Limit length
}

// Helper to create directory if it doesn't exist
function ensureDirectoryExists(dirPath) {
  if (!fs.existsSync(dirPath)) {
    fs.mkdirSync(dirPath, { recursive: true });
  }
}

// Helper to encode URL properly
function encodeUrl(url) {
  return encodeURIComponent(url).replace(/[!'()*]/g, function(c) {
    return '%' + c.charCodeAt(0).toString(16);
  });
}

// Generate product static HTML files
function generateProductFiles() {
  console.log('Generating product static HTML files...');
  
  // Load product data
  const products = JSON.parse(fs.readFileSync('./src/app/product/products_list_mk.json', 'utf8'));
  
  // Create products directory in root
  const productsDir = './static-products';
  ensureDirectoryExists(productsDir);

  products.forEach(product => {
    const id = product.id;
    const name = escapeHtml(product.name);
    const description = escapeHtml(product.description[0] || '');
    
    // Get the first product image or use default
    let imageUrl = DOMAIN + 'assets/favorit-logo.png';
    if (product.pictures && product.pictures[0]) {
      // Clean up the image path and encode it properly
      const imagePath = product.pictures[0].replace(/ /g, '%20');
      imageUrl = DOMAIN + imagePath;
    }

    const url = `${DOMAIN}product/${id}`;
    const currentUrl = `${DOMAIN}new/p/${id}?lang=mk`;

    const html = `<!DOCTYPE html>
<html lang="mk">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>${name} - Favorit Electronics</title>
  
  <!-- Open Graph Meta Tags -->
  <meta property="og:title" content="${name}">
  <meta property="og:description" content="${description}">
  <meta property="og:image" content="${imageUrl}">
  <meta property="og:image:width" content="1200">
  <meta property="og:image:height" content="630">
  <meta property="og:image:alt" content="${name}">
  <meta property="og:type" content="product">
  <meta property="og:url" content="${currentUrl}">
  <meta property="og:site_name" content="Favorit Electronics">
  <meta property="og:locale" content="mk">
  <meta property="og:locale:alternate" content="en">
  <meta property="og:locale:alternate" content="sr">
  <meta property="og:locale:alternate" content="al">
  
  <!-- Twitter Card Meta Tags -->
  <meta name="twitter:card" content="summary_large_image">
  <meta name="twitter:title" content="${name}">
  <meta name="twitter:description" content="${description}">
  <meta name="twitter:image" content="${imageUrl}">
  <meta name="twitter:image:alt" content="${name}">
  <meta name="twitter:site" content="@favoritelectronics">
  
  <!-- WhatsApp and Viber Meta Tags -->
  <meta property="og:image:type" content="image/png">
  <meta property="og:image:secure_url" content="${imageUrl}">
  
  <!-- SEO Meta Tags -->
  <meta name="description" content="${description}">
  <meta name="keywords" content="${name}, Favorit Electronics, home appliances, ${product.category || 'electronics'}, Macedonia">
  <meta name="author" content="Favorit Electronics">
  <meta name="robots" content="index, follow">
  <meta name="googlebot" content="index, follow">
  <meta name="language" content="Macedonian">
  <meta name="revisit-after" content="7 days">
  
  <!-- Canonical URL -->
  <link rel="canonical" href="${currentUrl}">
  
  <!-- Favicon -->
  <link rel="icon" type="image/x-icon" href="${DOMAIN}favicon.ico">
  
  <!-- Structured Data for Products -->
  <script type="application/ld+json">
  {
    "@context": "https://schema.org",
    "@type": "Product",
    "name": "${name}",
    "description": "${description}",
    "image": "${imageUrl}",
    "brand": {
      "@type": "Brand",
      "name": "Favorit Electronics"
    },
    "offers": {
      "@type": "Offer",
      "availability": "https://schema.org/InStock",
      "seller": {
        "@type": "Organization",
        "name": "Favorit Electronics"
      }
    }
  }
  </script>
  
  <style>
    body {
      font-family: Arial, sans-serif;
      margin: 0;
      padding: 20px;
      background: #f5f5f5;
      text-align: center;
    }
    .container {
      max-width: 600px;
      margin: 0 auto;
      background: white;
      padding: 40px;
      border-radius: 10px;
      box-shadow: 0 2px 10px rgba(0,0,0,0.1);
    }
    .logo {
      width: 200px;
      margin-bottom: 30px;
    }
    .product-image {
      max-width: 100%;
      height: auto;
      border-radius: 8px;
      margin: 20px 0;
      box-shadow: 0 4px 8px rgba(0,0,0,0.1);
    }
    h1 {
      color: #204880;
      margin-bottom: 20px;
      font-size: 24px;
    }
    p {
      color: #666;
      line-height: 1.6;
      margin-bottom: 30px;
      font-size: 16px;
    }
    .btn {
      background: #204880;
      color: white;
      padding: 15px 30px;
      text-decoration: none;
      border-radius: 5px;
      display: inline-block;
      margin: 10px;
      transition: background 0.3s;
      font-weight: bold;
    }
    .btn:hover {
      background: #2c5aa0;
    }
    .redirect-info {
      background: #e8f4fd;
      padding: 20px;
      border-radius: 5px;
      margin-top: 20px;
      color: #204880;
    }
    .meta-info {
      background: #f8f9fa;
      padding: 15px;
      border-radius: 5px;
      margin: 20px 0;
      font-size: 14px;
      color: #666;
    }
  </style>
</head>
<body>
  <div class="container">
    <img src="${DOMAIN}assets/favorit-logo.png" alt="Favorit Electronics" class="logo">
    
    <img src="${imageUrl}" alt="${name}" class="product-image">
    
    <h1>${name}</h1>
    <p>${description}</p>
    
    <div class="meta-info">
      <strong>Product ID:</strong> ${id}<br>
      <strong>Category:</strong> Favorit Electronics<br>
      <strong>Available:</strong> In Stock
    </div>
    
    <div class="redirect-info">
      <p>Redirecting to the full product page...</p>
    </div>
    
    <a href="${url}" class="btn">View Full Product</a>
    <a href="${DOMAIN}" class="btn">Back to Home</a>
  </div>
  
  <script>
    // Redirect after 5 seconds to allow crawlers to read meta tags
    setTimeout(function() {
      window.location.href = "${url}";
    }, 5000);
  </script>
</body>
</html>`;

    // Save to static-products directory
    fs.writeFileSync(`${productsDir}/product-${id}.html`, html, 'utf8');
    
    console.log(`Generated product-${id}.html`);
  });
}

// Generate blog static HTML files
function generateBlogFiles() {
  console.log('Generating blog static HTML files...');
  
  // Load blog data for all languages
  const blogs_mk = JSON.parse(fs.readFileSync('./src/app/blog-list/blogs_mk.json', 'utf8'));
  const blogs_en = JSON.parse(fs.readFileSync('./src/app/blog-list/blogs_en.json', 'utf8'));
  const blogs_sr = JSON.parse(fs.readFileSync('./src/app/blog-list/blogs_sr.json', 'utf8'));
  const blogs_al = JSON.parse(fs.readFileSync('./src/app/blog-list/blogs_al.json', 'utf8'));
  
  // Create blogs directory in root
  const blogsDir = './static-blogs';
  ensureDirectoryExists(blogsDir);

  // Process blogs for each language
  const languages = [
    { code: 'mk', data: blogs_mk.blogs },
    { code: 'en', data: blogs_en.blogs },
    { code: 'sr', data: blogs_sr.blogs },
    { code: 'al', data: blogs_al.blogs }
  ];

  languages.forEach(lang => {
    const langDir = `${blogsDir}/${lang.code}`;
    ensureDirectoryExists(langDir);

    lang.data.forEach(blog => {
      const id = blog.id;
      const sanitizedId = sanitizeFilename(id);
      const title = escapeHtml(blog.title);
      const excerpt = escapeHtml(blog.excerpt);
      const image = blog.image ? DOMAIN + blog.image : DOMAIN + 'assets/favorit-logo.png';
      const url = `${DOMAIN}blog/${id}?lang=${lang.code}`;
      const currentUrl = `${DOMAIN}new/static-blogs/${lang.code}/blog-${sanitizedId}.html`;
      const publishDate = blog.date || new Date().toISOString().split('T')[0];

      const html = `<!DOCTYPE html>
<html lang="${lang.code}">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>${title} - Favorit Electronics</title>
  
  <!-- Open Graph Meta Tags -->
  <meta property="og:title" content="${title}">
  <meta property="og:description" content="${excerpt}">
  <meta property="og:image" content="${image}">
  <meta property="og:image:width" content="1200">
  <meta property="og:image:height" content="630">
  <meta property="og:image:alt" content="${title}">
  <meta property="og:type" content="article">
  <meta property="og:url" content="${currentUrl}">
  <meta property="og:site_name" content="Favorit Electronics">
  <meta property="og:locale" content="${lang.code === 'mk' ? 'mk_MK' : lang.code === 'en' ? 'en_US' : lang.code === 'sr' ? 'sr_RS' : 'sq_AL'}">
  <meta property="article:published_time" content="${publishDate}">
  <meta property="article:author" content="Favorit Electronics">
  <meta property="article:section" content="Blog">
  
  <!-- Twitter Card Meta Tags -->
  <meta name="twitter:card" content="summary_large_image">
  <meta name="twitter:title" content="${title}">
  <meta name="twitter:description" content="${excerpt}">
  <meta name="twitter:image" content="${image}">
  <meta name="twitter:image:alt" content="${title}">
  <meta name="twitter:site" content="@favoritelectronics">
  
  <!-- Additional Meta Tags -->
  <meta name="description" content="${excerpt}">
  <meta name="keywords" content="Favorit Electronics, blog, ${title}, household appliances, Macedonia">
  <meta name="author" content="Favorit Electronics">
  <meta name="robots" content="index, follow">
  
  <!-- Canonical URL -->
  <link rel="canonical" href="${url}">
  
  <!-- Favicon -->
  <link rel="icon" type="image/x-icon" href="${DOMAIN}favicon.ico">
  
  <!-- Structured Data for Articles -->
  <script type="application/ld+json">
  {
    "@context": "https://schema.org",
    "@type": "Article",
    "headline": "${title}",
    "description": "${excerpt}",
    "image": "${image}",
    "author": {
      "@type": "Organization",
      "name": "Favorit Electronics"
    },
    "publisher": {
      "@type": "Organization",
      "name": "Favorit Electronics",
      "logo": {
        "@type": "ImageObject",
        "url": "${DOMAIN}assets/favorit-logo.png"
      }
    },
    "datePublished": "${publishDate}",
    "dateModified": "${publishDate}"
  }
  </script>
  
  <style>
    body {
      font-family: Arial, sans-serif;
      margin: 0;
      padding: 20px;
      background: #f5f5f5;
      text-align: center;
    }
    .container {
      max-width: 600px;
      margin: 0 auto;
      background: white;
      padding: 40px;
      border-radius: 10px;
      box-shadow: 0 2px 10px rgba(0,0,0,0.1);
    }
    .logo {
      width: 200px;
      margin-bottom: 30px;
    }
    .blog-image {
      max-width: 100%;
      height: auto;
      border-radius: 8px;
      margin: 20px 0;
      box-shadow: 0 4px 8px rgba(0,0,0,0.1);
    }
    h1 {
      color: #204880;
      margin-bottom: 20px;
      font-size: 24px;
    }
    p {
      color: #666;
      line-height: 1.6;
      margin-bottom: 30px;
      font-size: 16px;
    }
    .btn {
      background: #204880;
      color: white;
      padding: 15px 30px;
      text-decoration: none;
      border-radius: 5px;
      display: inline-block;
      margin: 10px;
      transition: background 0.3s;
      font-weight: bold;
    }
    .btn:hover {
      background: #2c5aa0;
    }
    .redirect-info {
      background: #e8f4fd;
      padding: 20px;
      border-radius: 5px;
      margin-top: 20px;
      color: #204880;
    }
    .meta-info {
      background: #f8f9fa;
      padding: 15px;
      border-radius: 5px;
      margin: 20px 0;
      font-size: 14px;
      color: #666;
    }
  </style>
</head>
<body>
  <div class="container">
    <img src="${DOMAIN}assets/favorit-logo.png" alt="Favorit Electronics" class="logo">
    
    <img src="${image}" alt="${title}" class="blog-image">
    
    <h1>${title}</h1>
    <p>${excerpt}</p>
    
    <div class="meta-info">
      <strong>Published:</strong> ${publishDate}<br>
      <strong>Category:</strong> Blog<br>
      <strong>Language:</strong> ${lang.code.toUpperCase()}
    </div>
    
    <div class="redirect-info">
      <p>Redirecting to the full blog post...</p>
    </div>
    
    <a href="${url}" class="btn">Read Full Article</a>
    <a href="${DOMAIN}" class="btn">Back to Home</a>
  </div>
  
  <script>
    // Redirect after 5 seconds to allow crawlers to read meta tags
    setTimeout(function() {
      window.location.href = "${url}";
    }, 5000);
  </script>
</body>
</html>`;

      fs.writeFileSync(`${langDir}/blog-${sanitizedId}.html`, html, 'utf8');
      console.log(`Generated blog-${sanitizedId}.html for language ${lang.code}`);
    });
  });
}

// Main execution
try {
  generateProductFiles();
  generateBlogFiles();
  console.log('Static HTML generation completed successfully!');
  console.log('Files generated in:');
  console.log('- ./static-products/ (optimized product files with SEO + Open Graph)');
  console.log('- ./static-blogs/ (blog files by language)');
} catch (error) {
  console.error('Error generating static HTML files:', error);
}

File: generate_static_html.py
#!/usr/bin/env python3
"""
Static HTML Generator for Favorit Electronics
Generates SEO-optimized static HTML files from JSON product data
"""

import json
import os
import re
from datetime import datetime
from urllib.parse import quote

class StaticHTMLGenerator:
    def __init__(self, base_url="https://favoritelectronics.com"):
        self.base_url = base_url
        self.languages = ['sr', 'mk', 'en', 'al']
        self.output_dir = "."
        
    def load_product_data(self, language):
        """Load product data for specific language"""
        file_path = f"src/app/product/products_list_{language}.json"
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                return json.load(f)
        except FileNotFoundError:
            print(f"Warning: {file_path} not found")
            return []
    
    def generate_meta_tags(self, product, language):
        """Generate comprehensive meta tags for social media"""
        
        # Clean and prepare data
        title = self.clean_text(product['name'])
        description = self.clean_text(product['description'][0]) if product['description'] else ""
        image_url = self.get_primary_image(product)
        product_url = f"{self.base_url}/product/{product['id']}"
        
        # Language-specific settings
        lang_codes = {'sr': 'sr', 'mk': 'mk', 'en': 'en', 'al': 'sq'}
        lang_code = lang_codes.get(language, 'en')
        
        meta_tags = f"""
  <meta charset="UTF-8">
  <title>{title} - Favorit Electronics</title>
  <meta name="description" content="{description[:160]}">
  <meta name="keywords" content="favorit, electronics, {self.extract_keywords(title)}">
  <meta name="robots" content="index, follow">
  <meta name="language" content="{lang_code}">
  <meta name="author" content="Favorit Electronics">
  <link rel="canonical" href="{product_url}">
  
  <!-- Open Graph Meta Tags -->
  <meta property="og:title" content="{title}">
  <meta property="og:description" content="{description[:200]}">
  <meta property="og:image" content="{image_url}">
  <meta property="og:image:width" content="1200">
  <meta property="og:image:height" content="630">
  <meta property="og:image:alt" content="{title}">
  <meta property="og:image:type" content="image/png">
  <meta property="og:type" content="product">
  <meta property="og:url" content="{product_url}">
  <meta property="og:site_name" content="Favorit Electronics">
  <meta property="og:locale" content="{self.get_og_locale(language)}">
  
  <!-- Twitter Card Meta Tags -->
  <meta name="twitter:card" content="summary_large_image">
  <meta name="twitter:title" content="{title}">
  <meta name="twitter:description" content="{description[:200]}">
  <meta name="twitter:image" content="{image_url}">
  <meta name="twitter:image:alt" content="{title}">
  <meta name="twitter:site" content="@favoritelectronics">
  <meta name="twitter:creator" content="@favoritelectronics">
  
  <!-- Additional SEO Meta Tags -->
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="theme-color" content="#1976d2">
  <meta name="msapplication-TileColor" content="#1976d2">
  
  <!-- Facebook App ID (replace with your actual Facebook App ID) -->
  <meta property="fb:app_id" content="YOUR_FACEBOOK_APP_ID">
  
  <!-- Structured Data (JSON-LD) -->
  <script type="application/ld+json">
  {{
    "@context": "https://schema.org",
    "@type": "Product",
    "name": "{title}",
    "description": "{description}",
    "image": "{image_url}",
    "brand": {{
      "@type": "Brand",
      "name": "Favorit Electronics"
    }},
    "manufacturer": {{
      "@type": "Organization",
      "name": "Favorit Electronics"
    }},
    "offers": {{
      "@type": "Offer",
      "url": "{product_url}",
      "availability": "https://schema.org/InStock"
    }}
  }}
  </script>"""
        
        return meta_tags
    
    def clean_text(self, text):
        """Clean text for meta tags"""
        if not text:
            return ""
        # Remove HTML tags
        text = re.sub(r'<[^>]+>', '', text)
        # Clean quotes and special characters
        text = text.replace('"', '&quot;').replace("'", '&#39;')
        # Remove extra whitespace
        text = ' '.join(text.split())
        return text
    
    def get_primary_image(self, product):
        """Get the primary product image URL"""
        if product.get('pictures') and len(product['pictures']) > 0:
            image_path = product['pictures'][0]
            # Ensure proper URL encoding
            encoded_path = quote(image_path, safe='/')
            return f"{self.base_url}/{encoded_path}"
        return f"{self.base_url}/assets/default-product.png"
    
    def extract_keywords(self, title):
        """Extract keywords from product title"""
        # Simple keyword extraction
        words = re.findall(r'\b\w+\b', title.lower())
        return ', '.join(words[:5])
    
    def get_og_locale(self, language):
        """Get Open Graph locale for language"""
        locales = {'sr': 'sr_RS', 'mk': 'mk_MK', 'en': 'en_US', 'al': 'sq_AL'}
        return locales.get(language, 'en_US')
    
    def get_macedonian_title(self, product):
        """Get Macedonian title for product"""
        mk_products = self.load_product_data('mk')
        for p in mk_products:
            if p['id'] == product['id']:
                return self.clean_text(p['name'])
        return self.clean_text(product['name'])
    
    def get_macedonian_description(self, product):
        """Get Macedonian description for product"""
        mk_products = self.load_product_data('mk')
        for p in mk_products:
            if p['id'] == product['id']:
                return self.clean_text(p['description'][0]) if p['description'] else ''
        return self.clean_text(product['description'][0]) if product['description'] else ''
    
    def get_serbian_title(self, product):
        """Get Serbian title for product"""
        sr_products = self.load_product_data('sr')
        for p in sr_products:
            if p['id'] == product['id']:
                return self.clean_text(p['name'])
        return self.clean_text(product['name'])
    
    def get_serbian_description(self, product):
        """Get Serbian description for product"""
        sr_products = self.load_product_data('sr')
        for p in sr_products:
            if p['id'] == product['id']:
                return self.clean_text(p['description'][0]) if p['description'] else ''
        return self.clean_text(product['description'][0]) if product['description'] else ''
    
    def get_english_title(self, product):
        """Get English title for product"""
        en_products = self.load_product_data('en')
        for p in en_products:
            if p['id'] == product['id']:
                return self.clean_text(p['name'])
        return self.clean_text(product['name'])
    
    def get_english_description(self, product):
        """Get English description for product"""
        en_products = self.load_product_data('en')
        for p in en_products:
            if p['id'] == product['id']:
                return self.clean_text(p['description'][0]) if p['description'] else ''
        return self.clean_text(product['description'][0]) if product['description'] else ''
    
    def get_albanian_title(self, product):
        """Get Albanian title for product"""
        al_products = self.load_product_data('al')
        for p in al_products:
            if p['id'] == product['id']:
                return self.clean_text(p['name'])
        return self.clean_text(product['name'])
    
    def get_albanian_description(self, product):
        """Get Albanian description for product"""
        al_products = self.load_product_data('al')
        for p in al_products:
            if p['id'] == product['id']:
                return self.clean_text(p['description'][0]) if p['description'] else ''
        return self.clean_text(product['description'][0]) if product['description'] else ''
    
    def generate_html_content(self, product, language):
        """Generate complete HTML content for a product"""
        
        meta_tags = self.generate_meta_tags(product, language)
        product_url = f"{self.base_url}/product/{product['id']}"
        
        html_content = f"""<!DOCTYPE html>
<html lang="{language}">
<head>
{meta_tags}
</head>
    <body>
      <script>
        // Redirect users to Angular app immediately
        setTimeout(() => {{
          window.location = "{product_url}?from_static=true";
        }}, 100);
      </script>
      <noscript>
        <!-- Fallback for users with JavaScript disabled -->
        <meta http-equiv="refresh" content="0; url={product_url}?from_static=true">
      </noscript>
      
      <!-- Fallback content for search engines -->
      <div style="display: none;">
        <h1>{product['name']}</h1>
        <p>{product['description'][0] if product['description'] else ''}</p>
        <img src="{self.get_primary_image(product)}" alt="{product['name']}">
      </div>
    </body>
</html>"""
        
        return html_content
    
    def generate_all_static_files(self):
        """Generate static HTML files for all products in Macedonian only"""
        
        print("Starting static HTML generation...")
        print("Processing mk language only...")
        
        # Only process Macedonian language
        language = 'mk'
        products = self.load_product_data(language)
        
        if not products:
            print(f"No products found for {language}")
            return
        
        for product in products:
            try:
                html_content = self.generate_html_content(product, language)
                filename = f"product-{product['id']}.html"
                
                with open(filename, 'w', encoding='utf-8') as f:
                    f.write(html_content)
                    
                print(f"Generated: {filename}")
                    
            except Exception as e:
                print(f"Error generating {filename}: {e}")
        
        print("Static HTML generation completed!")
    
    def generate_multi_language_html(self, product_id, language_data):
        """Generate HTML with language detection"""
        
        # Get default language data
        default_lang = list(language_data.keys())[0]
        default_product = language_data[default_lang]
        
        # Create language-specific meta tags
        meta_tags_by_lang = {}
        for lang, product in language_data.items():
            meta_tags_by_lang[lang] = self.generate_meta_tags(product, lang)
        
        # Generate JavaScript for language detection
        lang_detection_js = self.generate_language_detection_js(product_id, language_data)
        
        html_content = f"""<!DOCTYPE html>
<html lang="{default_lang}">
<head>
{meta_tags_by_lang[default_lang]}
</head>
<body>
  <script>
    {lang_detection_js}
  </script>
  <noscript>
    <!-- Fallback for users with JavaScript disabled -->
    <meta http-equiv="refresh" content="0; url={self.base_url}/product/{product_id}?lang={default_lang}&from_static=true">
  </noscript>
  
  <!-- Fallback content for search engines -->
  <div style="display: none;">
    <h1>{default_product['name']}</h1>
    <p>{default_product['description'][0] if default_product['description'] else ''}</p>
    <img src="{self.get_primary_image(default_product)}" alt="{default_product['name']}">
  </div>
</body>
</html>"""
        
        return html_content
    
    def generate_language_detection_js(self, product_id, language_data):
        """Generate JavaScript for language detection and meta tag updates"""
        
        js_code = f"""
    // Language detection and meta tag updates
    const urlParams = new URLSearchParams(window.location.search);
    const requestedLang = urlParams.get('lang') || 'mk';
    const supportedLangs = {list(language_data.keys())};
    const lang = supportedLangs.includes(requestedLang) ? requestedLang : 'mk';
    
    // Update meta tags based on language
    const metaData = {self.get_meta_data_json(language_data)};
    
    if (metaData[lang]) {{
      const data = metaData[lang];
      
      // Update title
      document.title = data.title;
      
      // Update meta tags
      document.querySelector('meta[name="description"]').setAttribute('content', data.description);
      document.querySelector('meta[property="og:title"]').setAttribute('content', data.title);
      document.querySelector('meta[property="og:description"]').setAttribute('content', data.description);
      document.querySelector('meta[property="og:url"]').setAttribute('content', data.url);
      document.querySelector('meta[property="og:locale"]').setAttribute('content', data.locale);
      document.querySelector('meta[name="twitter:title"]').setAttribute('content', data.title);
      document.querySelector('meta[name="twitter:description"]').setAttribute('content', data.description);
      
      // Update hidden content
      document.querySelector('h1').textContent = data.title;
      document.querySelector('p').textContent = data.description;
      document.querySelector('img').setAttribute('alt', data.title);
    }}
    
    // Redirect to Angular app
    window.location = '{self.base_url}/product/{product_id}?lang=' + lang + '&from_static=true';
"""
        return js_code
    
    def get_meta_data_json(self, language_data):
        """Get meta data as JSON for JavaScript"""
        import json
        meta_data = {}
        
        for lang, product in language_data.items():
            meta_data[lang] = {
                'title': self.clean_text(product['name']),
                'description': self.clean_text(product['description'][0]) if product['description'] else '',
                'url': f"{self.base_url}/product/{product['id']}",
                'locale': self.get_og_locale(lang)
            }
        
        return json.dumps(meta_data, ensure_ascii=False)

# Usage
if __name__ == "__main__":
    generator = StaticHTMLGenerator()
    generator.generate_all_static_files()

